kimchi/src/prover.rs

```diff
@@ -45,7 +44,43 @@ use rayon::iter::{
 use std::collections::HashMap;
 
 /// The result of a proof creation or verification.
-pub type Result<T> = std::result::Result<T, ProofError>;
+type Result<T> = std::result::Result<T, ProofError>;
+
+/// Contains variables needed for lookup in the prover algorithm.
+#[derive(Default)]
+struct LookupContext<G, F>
+where
+    G: CommitmentCurve,
+    F: FftField,
+{
+    /// The joint combiner used to join the columns of lookup tables
+    joint_combiner: Option<F>,
```

Comment:
It would be nice to use the type system to validate the transitions, rather than this runtime setup. I think we discussed using a trait like
```rust
trait LookupState {
  type JointCombiner;
   ...
}
struct LookupContext<State: LookupState> {
  joint_combiner: State::JointCombiner,
  ...
}
```
and then the helper functions in lookup can consume and construct the right contexts.

Reply:
yeah this is just the simplest way to move the lookup logic into optional blocks, we definitely should discuss more on how we can make the state transitions nicer :o